x/validator-preference/README.md

```diff
@@ -0,0 +1,155 @@
+- Give users a single message to stake {X} tokens, according to their validator-set preference distribution.
+- Give users a single message to unstake {X} tokens, according to their validator-set preference distribution.
+- Give users a single message to claim rewards from everyone on their preference list.
+- If a user does not have a validator-set preference list set and has is attempting to stake do the following; 
+  - Make their preference list default to their current staking distribution.
+- If a user has no preference list and no staking, then return error for messages.
+
+## Calculations
+
+Staking Calculation 
+
+- The user provides an amount to stake and our `MsgStakeToValidatorSet` divides the amount based on validator weight distribution.
+  For example: Stake 100osmo with validator-set {ValA -> 0.5, ValB -> 0.3, ValC -> 0.2}
+  our stake logic will attempt to stake (100 * 0.5) 50osmo for ValA , (100 * 0.3) 30osmo from ValB and (100 * 0.2) 20osmo from ValC.
+
+UnStaking Calculation 
+
+- The user provides an amount to unstake and our `MsgUnStakeFromValidatorSet` divides the amount based on validator weight distribution.
+- Here, the user can either unstake the entire amount or partial amount 
+  - Entire amount unstaking: UnStake 100osmo from validator-set {ValA -> 0.5, ValB -> 0.3, ValC -> 0.2},
+    our unstake logic will attempt to unstake 50osmo from ValA , 30osmo from ValB, 20osmo from ValC
+  - Partial amount unstaking: UnStake 27osmo from validator-set {ValA -> 0.5, ValB -> 0.3, ValC -> 0.2}, 
+    our unstake logic will attempt to unstake (27 * 0.5) 13.5osmos from ValA, (27 * 0.3), 8.1osmo from ValB, 
+    and (50 * 0.2) 5.4smo from ValC where 13.5osmo + 8.1osmo + 5.4osmo = 27osmo
+  - The user will then have 73osmo remaining with unchanged weights {ValA -> 0.5, ValB -> 0.3, ValC -> 0.2},
+
+## Messages
+
+### CreateValidatorSetPreference
+
+Creates a validator-set of `{valAddr, Weight}` given the delegator address.
+and preferences. The weights are in decimal format from 0 to 1 and must add up to 1.
+
+```go
+    string delegator = 1 [ (gogoproto.moretags) = "yaml:\"delegator\"" ];
+    repeated ValidatorPreference preferences = 2 [
+      (gogoproto.moretags) = "yaml:\"preferences\"",
+      (gogoproto.nullable) = false
+    ];
+```
+
+**State Modifications:**
+
+- Safety Checks
+  - check if the user already has a validator-set created. 
+  - check if the validator exist and is valid.
+  - check if the validator-set add up to 1.
+- Add owner address to the `KVStore`, where a state of validator-set is stored. 
+
+### UpdateValidatorSetPreference
+
+Updates a validator-set of `{valAddr, Weight}` given the delegator address
+and existing preferences. The weights calculations follow the same rule as `CreateValidatorSetPreference`.
+If a user changes their preferences list, the unstaking logic will run from the old set and 
+restaking to a new set is going to happen behind the scenes.
```

Comment:
Interesting!

What I'd suggest as the abstraction is we have the following messages:
* SetValidatorSet
* DelegateToValidatorSet
* UndelegateFromValidatorSet
* RedelegateToValidatorSet

So when I do "set", I don't adjust existing delegations. We only adjust existing delegations, when we do redelegate.

Reply:
(we can also implement modularly)

Reply:
This makes sense. `RedelegateToValidatorSet` will basically have the same features as `UpdateValidatorSetPreference` right? 

I like the option of supporting both set + redelegate function to make it modular. The only thing to check will be if the user already has a val-set. If they do then run  Regelegate logic else run SetValidator logic